test/cql-pytest/cassandra_tests/validation/operations/create_test.py

```diff
@@ -293,18 +293,18 @@ def testKeyspace(cql):
     n = unique_name()
     assertInvalidThrow(cql, n, SyntaxException, "CREATE KEYSPACE %s testXYZ ")
 
-    execute(cql, n, "CREATE KEYSPACE %s WITH replication = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 }")
+    execute(cql, n, "CREATE KEYSPACE %s WITH replication = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : 1 }")
```

Comment:
Does this change means, that if we start scylla with `--enable-tablets` and try to create SimpleStrategy it will try to turn tablets ON behind the scenes and then would fail to create SimpleStrategy because it's not tablet-aware?

Reply:
> it will try to turn tablets ON behind the scenes and then would fail to create SimpleStrategy because it's not tablet-aware?

If you mean that SimpleStrategy will try to turn tablets on and then will fail, it's not correct, the way I understand it, because we'd fail in the (query/strategy) validation phase, but then again, I'm not entirely sure what would mean for a strategy to turn tablets on - instantiate some tablets, maybe? If so, then no, because there're no tables in the ks at this points, AFAIK just defining a KS cannot result in creating any tablets. The code I changed will make this scenario look as if we're trying to create a KS with SimpleStrategy, and this simplestrategy C++ object carries a marker that it's tablets-enabled, which will immediately result in throwing an exception (n the validation phase), so that we won't do anything meaningful in the meantime.

Reply:
No, I mean -- why did you replace SimpleStrategy with NetworkTopologyStrategy in this test? It's generic cassandra test that should not depend on any tablets manipulations we're doing. Did it break after this patch/PR?

Reply:
Yes, this testcase did brake. I'm not allowing to create a ks with SimpleStrategy anymore, and since tests run with tablets enabled, this testcase tried to do exactly that: create a ks with simplestrategy with tablets. So I could approach this failing test in a couple of ways:
1. skip this test when tablets are enabled, but why skipping a good test when tablets are enabled? It can as well run, but only repl strategy needs to be changed for it to succeed,
2. explicitly specify that tablets are disabled in this ks by appending ` AND TABLETS = {'enable': false}`, but I had the same explanation as above,
3. change repl strategy from simple to network, so that the test runs and passes for both: tablets enabled and disabled. This is what I stayed with, because it still tests both scenarios (tablets enabled and disabled).

Reply:
>  I'm not allowing to create a ks with SimpleStrategy anymore

I think that's severe backward compatibility breakage. Even if we enable tablets "by default" it doesn't mean that strategies that don't support them start failing with default options. The correct logic is

```
    create_keyspace() {
        if (strategy.supports_tablets) {
            if (options.with_tablets == on) {
                if (cfg.enable_tablets) {
                    return create_keyspace_with_tablets();
                } else {
                    throw "tablets are not enabled";
                }
            } else if (options.with_tablets = off) {
                 return create_keyspace_without_tablets();
            } else { // options.with_tablets is not specified
                 if (cfg.enable_tablets) {
                     return create_keyspace_with_tablets();
                 } else {
                     return create_keyspace_without_tablets();
                 }
            }
        } else { // strategy doesn't support tablets
            if (options.with_tablets == on) {
                throw "invalid option";
            } else if (options.with_tablets == off) {
                throw "invalid option"; // but could as well return create_keyspace_without_tablets();
            } else { // options.with_tablets is not specified
                return create_keyspace_without_tablets();
            }
        }
    }
```

Reply:
@xemul in the `// strategy doesn't support tablets` you don't consider what `cfg.enable_tablets` is set to, and I somewhat understand this approach (was also considering all the options here https://github.com/scylladb/scylladb/issues/19744#issuecomment-2278296722), but `enable_tablets` from `scylla.yaml` says:

> Enables the tablets feature.
> **When enabled, newly created keyspaces will have tablets enabled by default**.
> That can be explicitly disabled in the CREATE KEYSPACE query
> by using the `tablets = {'enabled': false}` replication option.

I think that especially the 2nd sentence is important here, and it's the reason why I did what I did: in this testcase, there's a newly created KS that doesn't explicitly disable tablets and tablets are turned on in the config file, which implies that this KS needs to have tablets enabled (but it can't, since it's simplestrategy). 
So IDK :)

Reply:
Good point :) Let's make this comment less restrictive then too

Reply:
OK, I'll make the `enable_tablets` comment less restrictive to keep the code backward compatible. CC @bhalevy @tgrabiec I think you may be interested in this.

Reply:
@xemul one clarification question, as I have some second thoughts about this change being not backward compatible. So since 6.0 tablets are not backward compatible with anything, because they are a new feature, and once someone switches to use tablets, (s)he can't go back to vnodes:
1. if switching to tablets is done via an upgrade (e.g. from 5.4 to 6.0, and 6.0 has `tablets_enabled = true` by default), then the worse that can happen is that some scripts on the OSS user's side will break as they won't be able to create new keyspaces with simplestrategies, so these scripts/tools will have to be adjusted or tablets will have to be turned off (and turned on only later, once the scripts are fixed). But if a KS with simplestrategy existed, it'll continue to exist after the upgrade. What I'm trying to highlight here is that scylla won't break, only someone's scripting around scylla may break.
2. someone doesn't upgrade but starts using version 6.0+, and then (s)he has tablets enabled by default and just won't be able to create invalid keyspaces. In this case either the scylla.yaml config needs to be changed or CQL queries need to specify if tablets are disabled/enabled for certain keyspaces.

Of course these 2 are just very simplified scenarios, but I just don't know what exactly is not backward compatible and what's the impact/severity of changes that renders something not backward compatible. Can you help me figure out this?

Reply:
Done as stated in https://github.com/scylladb/scylladb/pull/20088#discussion_r1727105840, resolving comment.